src/cudamatrix/cu-matrix.h

```diff
@@ -334,6 +334,18 @@ class CuMatrixBase {
   void DiffLogSoftmaxPerRow(const CuMatrixBase<Real> &out_value,
                             const CuMatrixBase<Real> &out_deriv);
 
+  /// Add spatial regularization term to *this (`out_deriv`).
+  /// The regularization derivative is calculated from `out_value`.
+  /// Each row of `out_value` is first viewed as a 16*n grid,
+  /// convoluted with the 3x3 kernel twice, scaled with `-scale`
+  /// and then multiplied by `out_deriv`.
+  /// *this = *this - scale * Filter (X) (Filter (X) out_value)
+  /// If regularization_sumsq is not NULL, the sum of the squared
+  /// regularization term will be calculated and stored.
+  void AddSpatialRegularizationDeriv(const CuMatrixBase<Real>& out_value,
```

Comment:
Before we commit this, can you please change this to make it a non-member function and move it to cu-math.{h,cc}?  You can change the 'this' to be a pointer argument after 'scale', and call it 'deriv'.
I think in the past we have allowed too many 'special-purpose' functions to clutter the interface of class CuMatrix.  [e.g. the STL's 'string' class did this, and that is widely considered to have been a bad idea.]